### ISSUE TITLE
[JS/JSX][Formatting] Wrong indentation in this very particular case

### ISSUE THREAD

**Issue by jamesst20** [Microsoft/vscode#37407]
The Javascript formatter is close to perfection, there is only this particular case where it seems to break.

Before formatting: https://i.imgur.com/JpzSCPS.png
After formatting: https://i.imgur.com/4sz8Wop.png
Extensions list (in case) : https://i.imgur.com/MMV0ZiU.png

- VSCode Version: 1.17.2
- OS Version: Mac OS 10.12.6

Steps to Reproduce:

1. Fix manually indentation
2. Format document

Thank you very much !! :)

**Comment by Spown** [comment]
what did you expected? JSX is not even a superset of JS.

**Comment by Spown** [comment]
https://marketplace.visualstudio.com/items?itemName=gogocrow.jsx-beautify

**Comment by jamesst20** [comment]
That formatter makes the code looks awful. The original formatter does an awesome job and follows all the ESLint convention 99% of the time. The only exception is my screenshot.

Can it be fixed?

Thank you

**Comment by Spown** [comment]
>Can it be fixed?

its not broken. js-formatter does what it was designed to do - format es-comliant code. JSX is not es-compliant. just because it still does it better in some cases (99% in your case) than a dedicated formatter doesn't mean it has become it's job.

**Comment by jamesst20** [comment]
So I am assuming improving it is not an option where only a very little tweak could make it perfectly usable for both kind of project?

**Comment by Spown** [comment]
According to FAQ, [VSC supports JSX natively](https://code.visualstudio.com/docs/languages/javascript#_common-questions). Actually I don't know whether this support extends to formatting. If yes then - your bug-report is valid, if not and you are trying to format JSX with JS-formatter - then everything I said is applicable and you are filing a bug report to a wrong place.

If the later is true - a valid JSX code is invalid JS code. "improving" the formatter in such a way would mean the formatter would treat invalid js code as valid and format it accordingly, potentially producing scrambled mess. if you introduce some random unrelated syntax formatting rules into a formatter designed to format a specific sintax - you are opening doors for unforeseen consequences.

Still the best way to go would be to find the formatter code in the VSC repository, fork it into an extension. and improve it accordingly. maybe together with https://github.com/LiShiSangZi.

**Comment by jamesst20** [comment]
To be honnest, I would be supprise that the formatter is that good with JSX if it wasn't intended for. I am pretty sure it's only a little bug. If you take a look at the screenshot, you will notice that even the first part of the condition formats properly.

`{this.props.readOnly ? ( //Good
                              <span>{value}</span>//Good
                            ) : ( //Good
                              <TextField value={value} type="number" onChange={this.handlePricesChange(width, height)} /> // Wrong
                            )}` // Wrong

**Comment by jamesst20** [comment]
In fact, formatted this way, it is only the inner second condition:

Manually formatted: https://i.imgur.com/5xNjrMF.png
After: https://i.imgur.com/qqLWZ6m.png

**Comment by mjbvz** [comment]
Our builtin formatter fully supports jsx. It seems that you are running into this known TypeScript bug: https://github.com/Microsoft/TypeScript/issues/14494

Closing as duplicate of https://github.com/Microsoft/TypeScript/issues/14494
